### ISSUE TITLE
Can't compile the files!

### ISSUE THREAD

**Issue by 2loopVoyvoda** [soundcrowd/soundcrowd#81]
I've followed the instructions, but unfortunately I can't make an APK compile, because of the following error:
_Task :app:checkDebugAarMetadata FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:checkDebugAarMetadata'.
> Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
   > Could not resolve com.tiefensuche.soundcrowd:plugin:0.1.
     Required by:
         project :app
      > Could not resolve com.tiefensuche.soundcrowd:plugin:0.1.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-plugin/com/tiefensuche/soundcrowd/plugin/0.1/plugin-0.1.pom'.
            > Username must not be null!
      > Could not resolve com.tiefensuche.soundcrowd:plugin:0.1.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-waveform/com/tiefensuche/soundcrowd/plugin/0.1/plugin-0.1.pom'.
            > Username must not be null!
   > Could not resolve com.tiefensuche.soundcrowd:waveform:1.0.
     Required by:
         project :app
      > Could not resolve com.tiefensuche.soundcrowd:waveform:1.0.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-plugin/com/tiefensuche/soundcrowd/waveform/1.0/waveform-1.0.pom'.
            > Username must not be null!
      > Could not resolve com.tiefensuche.soundcrowd:waveform:1.0.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-waveform/com/tiefensuche/soundcrowd/waveform/1.0/waveform-1.0.pom'.
            > Username must not be null!_
So, as you can tell, my question is.. **What now? Any ideas?**

**Comment by 0xf4b1** [comment]
Hey, the build uses two artifacts that are deployed in the maven repository on github and it requires authentication even for reading and downloading them. You can create a personal token by following the instructions [here](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-personal-access-token-classic) and then adjust the `username` and `password` fields here: https://github.com/soundcrowd/soundcrowd/blob/da1c565c25822943e1bd86a1c5f1be1e04de4885/app/build.gradle#L70-L83

**Comment by IzzySoft** [comment]
@0xf4b1 that's not all. Your build (attached to the releases) has a `lib/arm64-v8a/libsongrec.so` that's not there if I build from sources. My build recipe, using `localMaven()` to build entirely from source (and not need those credentials):

```yaml
build:
  - git clone https://github.com/soundcrowd/soundcrowd-plugin.git
  - pushd soundcrowd-plugin
  - git checkout 5ef111cd40dd595c8831854da0106a37fa750784
  - ./gradlew publishToMavenLocal
  - popd
  - git clone https://github.com/soundcrowd/soundcrowd-waveform.git
  - pushd soundcrowd-waveform
  - git checkout 2d77eef1400d2496e901c174e405fec1d22eed38
  - ./gradlew publishToMavenLocal
  - popd
  - sed -r 's/mavenCentral\(\)/mavenCentral()\nmavenLocal()/' -i app/build.gradle
  - chmod +x gradlew
  - ./gradlew assembleRelease
```

succeeds to produce an APK, but that's about 1 MB smaller than yours, coming with the following differences:

```diff
- -rw-r--r--  0.0 unx      120 b-      116 defN 1981-01-01 01:01:02 6b7efe6a META-INF/version-control-info.textproto
- -rw-r--r--  0.0 unx     1419 b-     1419 stor 1981-01-01 01:01:02 a9d5ca12 assets/dexopt/baseline.prof
- -rw-r--r--  0.0 unx      238 b-      238 stor 1981-01-01 01:01:02 9d04ddfa assets/dexopt/baseline.profm
- -rw-r--r--  0.0 unx  7362768 b-  2827906 defN 1981-01-01 01:01:02 53b85934 classes.dex
- -rw-r--r--  0.0 unx  2273000 b-   938328 defN 1981-01-01 01:01:02 6f1b59ee lib/arm64-v8a/libsongrec.so
+ -rw-r--r--  0.0 unx      120 b-      117 defN 1981-01-01 01:01:02 5c766408 META-INF/version-control-info.textproto
+ -rw-r--r--  0.0 unx     1416 b-     1416 stor 1981-01-01 01:01:02 798a57da assets/dexopt/baseline.prof
+ -rw-r--r--  0.0 unx      237 b-      237 stor 1981-01-01 01:01:02 10688aec assets/dexopt/baseline.profm
+ -rw-r--r--  0.0 unx  7357628 b-  2825219 defN 1981-01-01 01:01:02 842ae5a5 classes.dex
  -rw----     2.0 fat     1738 b-      782 defN 1981-01-01 01:01:02 7cded4df DebugProbesKt.bin
  -rw----     2.0 fat        6 b-        8 defN 1981-01-01 01:01:02 38ca8087 META-INF/androidx.activity_activity.version
  -rw----     2.0 fat        6 b-        8 defN 1981-01-01 01:01:02 f6d036c7 META-INF/androidx.annotation_annotation-experimental.version
@@ -76,7 +75,7 @@
  -rw----     2.0 fat    18661 b-     5257 defN 1981-01-01 01:01:02 b623e8f5 kotlin/kotlin.kotlin_builtins
  -rw----     2.0 fat     3399 b-     1292 defN 1981-01-01 01:01:02 cb3b8aed kotlin/ranges/ranges.kotlin_builtins
  -rw----     2.0 fat     2426 b-     1286 defN 1981-01-01 01:01:02 7e3e3d18 kotlin/reflect/reflect.kotlin_builtins
- -rw----     0.0 fat    11300 b-     2748 defN 1981-01-01 01:01:02 0f157ad3 AndroidManifest.xml
+ -rw----     0.0 fat    11872 b-     2867 defN 1981-01-01 01:01:02 1d4dd0b0 AndroidManifest.xml
  -rw----     0.0 fat      744 b-      341 defN 1981-01-01 01:01:02 ce3ab127 res/-1.xml
  -rw----     0.0 fat      352 b-      207 defN 1981-01-01 01:01:02 51c8be78 res/-5.xml
  -rw----     0.0 fat      712 b-      377 defN 1981-01-01 01:01:02 f3a0795e res/-7.xml
@@ -176,7 +175,6 @@
  -rw----     0.0 fat      164 b-      164 stor 1981-01-01 01:01:02 f06b3df7 res/6U.png
  -rw----     0.0 fat      916 b-      408 defN 1981-01-01 01:01:02 6de82afb res/6f.xml
  -rw----     0.0 fat      227 b-      227 stor 1981-01-01 01:01:02 55e57039 res/6t.png
- -rw----     0.0 fat      584 b-      292 defN 1981-01-01 01:01:02 dd3883c4 res/6u.xml
  -rw----     0.0 fat      436 b-      260 defN 1981-01-01 01:01:02 66c3f4ea res/6x.xml
  -rw----     0.0 fat      430 b-      430 stor 1981-01-01 01:01:02 c7c65396 res/7C.9.png
  -rw----     0.0 fat     7870 b-     7870 stor 1981-01-01 01:01:02 a62679e1 res/7G.png
@@ -1139,8 +1137,5 @@
  -rw----     0.0 fat      464 b-      269 defN 1981-01-01 01:01:02 28c670a9 res/zq.xml
  -rw----     0.0 fat      682 b-      682 stor 1981-01-01 01:01:02 663fd684 res/zr.png
  -rw----     0.0 fat     2756 b-      953 defN 1981-01-01 01:01:02 fa5593cc res/zy.xml
- -rw----     0.0 fat   950524 b-   950524 stor 1981-01-01 01:01:02 818eedd4 resources.arsc
+ -rw----     0.0 fat   950468 b-   950468 stor 1981-01-01 01:01:02 e857c40d resources.arsc
```

(ignoring the signature here). Can you tell me what step I'm missing there? For some background: SoundCrowd is [listed at IzzyOnDroid](https://apt.izzysoft.de/packages/com.tiefensuche.soundcrowd), where we support [Reproducible Builds](https://reproducible-builds.org/) (see: [Reproducible Builds, special client support and more at IzzyOnDroid](https://android.izzysoft.de/articles/named/iod-rbs-mirrors-clients)), so I'm trying to add SoundCroud to the RB apps here. Would be great if you'd help us succeed with that. Thanks in advance!

PS: the APK attached to your release was obviously not built from the commit the tag points to (3cf406a6180a717b5c6769999c5848bbf805e98e), but from 242be6e92942013eb8088f1037e8dd013d611efe – which is in a different branch named `songrec`, which explains the missing lib (and probably some other differences) – and makes maintaining RB harder for those running the verification builders (as they have to identify the correct commit & branch first). Now, building from that branch fails as it cannot resolve `io.github.tiefensuche:songrec-android:0.1` – and I cannot find the corresponding repo in your account here. Where would I build that from?

**Comment by 0xf4b1** [comment]
Hey, yes this is correct, I built the attached file from the songrec branch which adds few commits on top of master and needs https://github.com/0xf4b1/songrec-android. I built from this branch only the artifact from releases as preview, it's not part of the official release in the playstore, and the songrec repo does not yet contain the needed build steps or scripts.

To avoid confusion, I've updated the `soundcrowd.apk` which is now cleanly built from master and the tagged version and I've renamed the other artifact to `soundcrowd-songrec.apk`, so you can ignore the songrec branch.

**Comment by IzzySoft** [comment]
Thanks! That compiles fine here, but the APKs are not identical. Diff of the APKs:

```diff
  -rw-r--r--  0.0 unx      120 b-      117 defN 1981-01-01 01:01:02 5c766408 META-INF/version-control-info.textproto
- -rw-r--r--  0.0 unx     1416 b-     1416 stor 1981-01-01 01:01:02 d2384f01 assets/dexopt/baseline.prof
+ -rw-r--r--  0.0 unx     1416 b-     1416 stor 1981-01-01 01:01:02 798a57da assets/dexopt/baseline.prof
  -rw-r--r--  0.0 unx      237 b-      237 stor 1981-01-01 01:01:02 10688aec assets/dexopt/baseline.profm
- -rw-r--r--  0.0 unx  7357628 b-  2825219 defN 1981-01-01 01:01:02 d99ea32d classes.dex
+ -rw-r--r--  0.0 unx  7357628 b-  2825219 defN 1981-01-01 01:01:02 842ae5a5 classes.dex
  -rw----     2.0 fat     1738 b-      782 defN 1981-01-01 01:01:02 7cded4df DebugProbesKt.bin
```

Baseline usually differs when `clsses.dex` differs, so let's take a ok at the latter. Lines are too long to render here in a useful way, so let me just attach the diff and point out the difference in words. At the very end of the line:

* yours: `k=1 mv={ 1 7 1 } xi=48`
* mine: `k=1 mv={ 1 8 0 } xi=48`

Almost looks like a version of something. Start of the line on both ends: `VISIBILITY_RUNTIME Lkotlin/Metadata; d1={` – not sure if that gives you a clue.

[dex.diff.txt](https://github.com/user-attachments/files/17970620/dex.diff.txt)

Had to rename the file to `.txt` as otherwise Github lets me not attach such "dangerous files" :stuck_out_tongue_winking_eye:

**TL;DR:** we're pretty close to a reproducible build here, it's "just" that single line…

PS: Better do not replace files which have already been distributed (too late now that both are out in the wild). The IzzyOnDroid repo now e.g. serves the file that had the same name before, but now is something different. With RBs active, such a rename would render the RB-results unusable :wink:

**Comment by IzzySoft** [comment]
Thanks! I've decided to build the artifacts from source for 2 reasons: 1) build failed for me if I didn't, and 2) speaking about open source builds and RB, it just feels right to do so :wink:

**Comment by 0xf4b1** [comment]
I absolutely agree with your strategy regarding reproduce builds to build everything from source if possible :)

I think I found the difference, it seems I've built the waveform module without https://github.com/soundcrowd/soundcrowd-waveform/commit/29afa12f75929e9778717cc86767a85c331398e5 and published, after reverting it and rebuilding waveform and then soundcrowd the diff is gone. I can take this commit out and properly tag the version in the repo, then it should be clear what was used the whole time. Actually you should have encountered the diff also in the previous versions if you did the same there already 🤔

**Comment by 0xf4b1** [comment]
@IzzySoft Please try to replace the waveform checkout `git checkout 2d77eef1400d2496e901c174e405fec1d22eed38` with  `git checkout 167dd57c3d387ce2ea106d9fdae55fd6ecec3a25` and I hope you'll no longer get a diff to the things I've released 🤞

**Comment by IzzySoft** [comment]
Right, no diffs. Not even an APK :see_no_evil:

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:checkReleaseAarMetadata'.
> Could not resolve all files for configuration ':app:releaseRuntimeClasspath'.
   > Could not resolve io.github.tiefensuche:songrec-android:0.1.
     Required by:
         project :app
      > Could not resolve io.github.tiefensuche:songrec-android:0.1.
         > Could not get resource 'https://jitpack.io/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.pom'.
            > Could not GET 'https://jitpack.io/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.pom'. Received status code 401 from server: Unauthorized
      > Could not resolve io.github.tiefensuche:songrec-android:0.1.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-plugin/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.pom'.
            > Username must not be null!
      > Could not resolve io.github.tiefensuche:songrec-android:0.1.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-waveform/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.pom'.
            > Username must not be null!
      > Could not resolve io.github.tiefensuche:songrec-android:0.1.
         > Could not get resource 'https://maven.pkg.github.com/0xf4b1/songrec-android/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.pom'.
            > Username must not be null!
```

did the tag move? I had that problem tackled last time, it should use local maven instead. Need to dig in what's up there, give me a little…

**Comment by 0xf4b1** [comment]
I was hoping to hear everything is good before I put it on master and tag 😅 

This error is because you built from songrec branch. I think what should be identical now is the updated soundcrowd.apk if you built the app from master (basically this small diff you had should be gone)

**Comment by IzzySoft** [comment]
Ah, songrec, there was something… There was a `git checkout 242be6e92942013eb8088f1037e8dd013d611efe` in my recipe for something(must have picked that from the versionInfo in the "other" APK), guess it was that. Removed it and the build is still running (eh, now our posts cross)… OK, builds – but is not RB. Eh, no diff? Wait, you're probably sign with…

    files with apksigner padding                : 313

Gna… OK, give me another 10 min, looks like we're almost there…

**Comment by IzzySoft** [comment]
```yaml
build:
  - git clone https://github.com/soundcrowd/soundcrowd-plugin.git
  - pushd soundcrowd-plugin
  - git checkout 5ef111cd40dd595c8831854da0106a37fa750784
  - ./gradlew publishToMavenLocal
  - popd
  - git clone https://github.com/soundcrowd/soundcrowd-waveform.git
  - pushd soundcrowd-waveform
  - git checkout 167dd57c3d387ce2ea106d9fdae55fd6ecec3a25
  - ./gradlew publishToMavenLocal
  - popd
  - sed -r 's/mavenCentral\(\)/mavenCentral()\nmavenLocal()/' -i app/build.gradle
  - chmod +x gradlew
  - ./gradlew assembleRelease
  - find . -name '*.apk'
  - git clone -b v0.3.0 https://github.com/obfusk/reproducible-apk-tools.git
  - reproducible-apk-tools/zipalign.py --page-size 4 --pad-like-apksigner --replace app/build/outputs/apk/release/app-release-unsigned.apk /outputs/unsigned.apk
```

because of

```
$ zipalignment *upstream.apk
  zipaligned (4-byte alignment)               : yes
  files with apksigner padding                : 313
  apksigner alignments from extra fields      : 4
  most likely uncompressed .so page alignment : none
```

:crossed_fingers: :crossed_fingers: :crossed_fingers: 

Yay! :partying_face: 

```
    "upstream_signed_apk_sha256": "e6608ae4b6c0e3bb853ff800dd7823ac29ccb8b1c11a27bf6d667fc0a0a94c0b",
    "built_unsigned_apk_sha256": "223da412e0f6b41dd828c7d315de0b2321fe4ae4d313c2782bdc025c87823cfa",
    "signature_copied_apk_sha256": "e6608ae4b6c0e3bb853ff800dd7823ac29ccb8b1c11a27bf6d667fc0a0a94c0b"
```

That means "RB" (first hash is yours, middle one my unsigned, last my unsigned with the signing copied from yours – match!) So let's roll that out. Some more thoughts in the next comment…

**Comment by 0xf4b1** [comment]
Very nice, happy that it worked out with the small change in waveform, updated now on master :)

Regarding submodules, I actually moved away from them [Remove plugin and waveform submodules](https://github.com/soundcrowd/soundcrowd/commit/c78bed32f2db577f50d14dd41d06e6ff9d9379a4) in favor of handling them same as any other dependency. The modules haven't changed for a long time and I don't have something in mind there, so I think you can keep it for now as it is or checkout the tag that equals the version instead of checking out a specific commit.

**Comment by IzzySoft** [comment]
If they rarely change, that should be OK. So we go with the "act when it breaks" variant then. Might be the technically ugliest one – but perfection is the enemy of the good, and all that :wink: Thanks!

**Comment by IzzySoft** [comment]
@0xf4b1 broke with 4.0.0. Using the very same recipe (ugh, what was that with "rarely change"?), I get

```
> Task :app:mergeReleaseNativeLibs
> Task :app:mergeReleaseNativeLibs FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeReleaseNativeLibs'.
> Could not resolve all files for configuration ':app:releaseRuntimeClasspath'.
   > Could not resolve com.tiefensuche.soundcrowd:plugin:1.0.0.
     Required by:
         project :app
         project :app > project :soundcloud
         project :app > project :spotify
         project :app > project :tidal
         project :app > project :youtube
      > Could not resolve com.tiefensuche.soundcrowd:plugin:1.0.0.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-plugin/com/tiefensuche/soundcrowd/plugin/1.0.0/plugin-1.0.0.pom'.
            > Username must not be null!
      > Could not resolve com.tiefensuche.soundcrowd:plugin:1.0.0.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-waveform/com/tiefensuche/soundcrowd/plugin/1.0.0/plugin-1.0.0.pom'.
            > Username must not be null!
      > Could not resolve com.tiefensuche.soundcrowd:plugin:1.0.0.
         > Could not get resource 'https://maven.pkg.github.com/0xf4b1/songrec-android/com/tiefensuche/soundcrowd/plugin/1.0.0/plugin-1.0.0.pom'.
            > Username must not be null!
   > Could not resolve io.github.tiefensuche:songrec-android:0.1.
     Required by:
         project :app
      > Could not resolve io.github.tiefensuche:songrec-android:0.1.
         > Could not get resource 'https://jitpack.io/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.pom'.
            > Could not GET 'https://jitpack.io/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.pom'. Received status code 401 from server: Un
      > Could not resolve io.github.tiefensuche:songrec-android:0.1.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-plugin/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.p
            > Username must not be null!
      > Could not resolve io.github.tiefensuche:songrec-android:0.1.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-waveform/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1
            > Username must not be null!
      > Could not resolve io.github.tiefensuche:songrec-android:0.1.
         > Could not get resource 'https://maven.pkg.github.com/0xf4b1/songrec-android/io/github/tiefensuche/songrec-android/0.1/songrec-android-0.1.pom'.
            > Username must not be null!
   > Could not resolve io.github.tiefensuche:spotify-kt:0.1.
     Required by:
         project :app > project :spotify
      > Could not resolve io.github.tiefensuche:spotify-kt:0.1.
         > Could not get resource 'https://jitpack.io/io/github/tiefensuche/spotify-kt/0.1/spotify-kt-0.1.pom'.
            > Could not GET 'https://jitpack.io/io/github/tiefensuche/spotify-kt/0.1/spotify-kt-0.1.pom'. Received status code 401 from server: Unauthorized
      > Could not resolve io.github.tiefensuche:spotify-kt:0.1.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-plugin/io/github/tiefensuche/spotify-kt/0.1/spotify-kt-0.1.pom'.
            > Username must not be null!
      > Could not resolve io.github.tiefensuche:spotify-kt:0.1.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-waveform/io/github/tiefensuche/spotify-kt/0.1/spotify-kt-0.1.pom'.
            > Username must not be null!
      > Could not resolve io.github.tiefensuche:spotify-kt:0.1.
         > Could not get resource 'https://maven.pkg.github.com/0xf4b1/songrec-android/io/github/tiefensuche/spotify-kt/0.1/spotify-kt-0.1.pom'.
            > Username must not be null!
   > Could not resolve com.tiefensuche:tidal-kt:0.2.0.
     Required by:
         project :app > project :tidal
      > Could not resolve com.tiefensuche:tidal-kt:0.2.0.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-plugin/com/tiefensuche/tidal-kt/0.2.0/tidal-kt-0.2.0.pom'.
            > Username must not be null!
      > Could not resolve com.tiefensuche:tidal-kt:0.2.0.
         > Could not get resource 'https://maven.pkg.github.com/soundcrowd/soundcrowd-waveform/com/tiefensuche/tidal-kt/0.2.0/tidal-kt-0.2.0.pom'.
            > Username must not be null!
      > Could not resolve com.tiefensuche:tidal-kt:0.2.0.
         > Could not get resource 'https://maven.pkg.github.com/0xf4b1/songrec-android/com/tiefensuche/tidal-kt/0.2.0/tidal-kt-0.2.0.pom'.
            > Username must not be null!
```

Hm, Github maven does not work without login, so that renders the app effectively non-RB.

I see there are several submodules. Though none of them are mentioned in your Github action, would iterating over them with `./gradlew publishToMavenLocal` help? And are the other two repos (soundcrowd-plugin and soundcrowd-waveform) still required to clone and also publish to mavenLocal?

**Comment by IzzySoft** [comment]
I've just checked, won't work: those are the ones requiring the credentials. So 4.0.0 won't be RB then. Or do you have any idea?

**Comment by 0xf4b1** [comment]
Just to clarify, the modules hosted on github are not private, they can even be downloaded via the website without being authenticated, but I know it is annoying that they can not be directly used by maven/gradle without a token of an arbitrary user.

The built itself is actually not that complex as it can be seen in the CI script, if the env variables GITHUB_ACTOR and GITHUB_TOKEN are given to pull from maven github repos.
Then its only a recursive clone to checkout the submodules followed by a ./gradlew assembleRelease

**Comment by 0xf4b1** [comment]
It is not only that single module, let me see if I can publish them elsewhere instead so that the build generally works without github tokens in the future

**Comment by IzzySoft** [comment]
Ah, cool! Please give me a ping when the next release is out so I give it another try!

**Comment by 0xf4b1** [comment]
Sure, I'll do this :)

**Comment by 0xf4b1** [comment]
@IzzySoft Hey, I just created a smaller new release, feel free to retry the build, it should work just with a recursive clone and gradle assembleRelease

**Comment by IzzySoft** [comment]
Almost. I also needed to match your "secrets" 😉 But then: yes, thanks, it built and is RB 🥳
